File: python/algos/sqrt_x.py
def my_sqrt(x: int) -> int:
    """Calculates integer sqrt."""
    if x == 0:
        return 0
    elif x < 4:
        return 1
    else:
        left = 2
        right = x // 2
        result = 2

        while left <= right:
            middle = left + (right - left) // 2
            if middle <= x / middle:
                left = middle + 1
                result = middle
            else:
                right = middle - 1

        return result


if __name__ == '__main__':
    assert my_sqrt(0) == 0
    assert my_sqrt(1) == 1
    assert my_sqrt(3) == 1
    assert my_sqrt(4) == 2
    assert my_sqrt(8) == 2
    assert my_sqrt(16) == 4
    assert my_sqrt(2147395599) == 46339

File: rust/src/algos/sqrt_x.rs
pub fn my_sqrt(x: i32) -> i32 {
    if x == 0 {
        0
    } else if x < 4 {
        1
    } else {
        let mut left = 1;
        let mut right = x / 2;
        let mut result = 0;

        while left <= right {
            let middle = left + (right - left) / 2;
            if middle <= x / middle {
                left = middle + 1;
                result = middle;
            } else {
                right = middle - 1;
            }
        }
        result
    }
}
